File: java/src/main/java/com/example/cache/RedisUtils.java
package com.example.cache;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;

import redis.clients.jedis.Jedis;

public class RedisUtils {
    // get does not distinguish "null" from empty
    public static <K extends Serializable> Function<K, Optional<Serializable>> get(Jedis cache) {
        return key -> Optional.ofNullable(cache.get(serialize(key))).map(RedisUtils::deserialize);
    }
    
    public static <K extends Serializable, V extends Serializable> BiConsumer<K, Optional<V>> put(Jedis cache) {
        return (key, value) -> {
            if (value.isPresent()) {
                cache.set(serialize(key), serialize(value.get()));
            } else {
                cache.del(serialize(key));
            }
        };
    }
    
    private static String serialize(Serializable a) {
        try {
            ByteArrayOutputStream bo = new ByteArrayOutputStream();
            ObjectOutputStream so = new ObjectOutputStream(bo);
            so.writeObject(a);
            so.flush();
            return bo.toString();
        } catch (IOException e) {
            return null; // TODO better error handling
        }
    }
    
    private static Serializable deserialize(String a) {
        try {
            byte b[] = a.getBytes(); 
            ByteArrayInputStream bi = new ByteArrayInputStream(b);
            ObjectInputStream si;
            si = new ObjectInputStream(bi);
            return (Serializable) si.readObject();
        } catch (IOException | ClassNotFoundException e) {
            return null; // TODO better error handling
        }
    }
}

File: scala/src/main/scala/com/example/cache/RedisUtils.scala
package com.example.cache

import java.util.function.BiConsumer
import redis.clients.jedis.Jedis
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectOutputStream
import java.io.ObjectInputStream
import java.io.ByteArrayInputStream

object RedisUtils {
    def get(cache: Jedis) = (key: Serializable) => {
        deserialize(cache.get(serialize(key).get))
    }: Option[Serializable]

    def put(cache: Jedis) = (key: Serializable, value: Serializable) => { 
        cache.set(serialize(key).get, serialize(value).get)
    }: Unit
    
    def serialize(a: Serializable) = {
        try {
            val bo = new ByteArrayOutputStream()
            val so = new ObjectOutputStream(bo)
            so.writeObject(a)
            so.flush()
            Some(bo.toString())
        } catch {
            case e: IOException => None   
        }
    }: Option[String]

    def deserialize(a: String) = {
        try {
            val b = a.getBytes() 
            val bi = new ByteArrayInputStream(b)
            val si = new ObjectInputStream(bi)
            Some(si.readObject().asInstanceOf[Serializable]);
        } catch {
            case e @ (_ : RuntimeException | _ : ClassNotFoundException) => None
        }
    }: Option[Serializable]
}
